python_modules/dagster/dagster/_core/definitions/external_asset.py

```diff
@@ -149,10 +162,64 @@ def create_external_asset_from_source_asset(source_asset: SourceAsset) -> Assets
         kwargs["io_manager_key"] = source_asset.io_manager_key
 
     @asset(**kwargs)
-    def _external_assets_def() -> None:
-        raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+    def _shim_assets_def(context: AssetExecutionContext):
+        if not source_asset.observe_fn:
+            raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+
+        op_function = wrap_source_asset_observe_fn_in_op_compute_fn(source_asset)
+        return_value = op_function.decorated_fn(context)
+        check.invariant(
+            return_value is None,
+            "The wrapped decorated_fn should return a value. If this changes, this code path must"
+            " changed to process the events appopriately.",
+        )
+
+    check.invariant(isinstance(_shim_assets_def, AssetsDefinition))
+    assert isinstance(_shim_assets_def, AssetsDefinition)  # appease pyright
+
+    return _shim_assets_def
 
-    check.invariant(isinstance(_external_assets_def, AssetsDefinition))
-    assert isinstance(_external_assets_def, AssetsDefinition)  # appese pyright
 
-    return _external_assets_def
+def create_observation_with_version(
+    asset_key: AssetKey, data_version: DataVersion
+) -> AssetObservation:
+    return AssetObservation(
+        asset_key=asset_key,
+        tags={DATA_VERSION_TAG: data_version.value},
+    )
+
+
+def get_auto_sensor_name(asset_key: AssetKey) -> str:
+    return f"__auto_observe_sensor{asset_key.to_python_identifier()}"
+
+
+def sensor_def_from_observable_source_asset(
+    observable_source_asset: SourceAsset,
+    sensor_name: Optional[str] = None,
+) -> SensorDefinition:
+    """Given an existing observable source asset, generate a sensor that observes it on a regular
+    interval as specified by `SourceAsset.auto_observe_internal_minutes`.
+    """
+    sensor_name = sensor_name if sensor_name else get_auto_sensor_name(observable_source_asset.key)
+
+    @sensor(
+        name=sensor_name,
+        minimum_interval_seconds=(
+            int(observable_source_asset.auto_observe_interval_minutes * 60)
+            if observable_source_asset.auto_observe_interval_minutes is not None
+            else 5
+            * 60  # I could not find the default value (undocumented) so guessing it is 5 minutes?
```

Comment:
@OwenKephart I was trying to find the default value for `auto_observe_interval_minutes`. I couldn't find it in the codebase. I'll update the docblock for `SourceAsset` once I know what it is.

Reply:
There's no default value for this (other than `None) -- if someone wants an observable source asset to be automatically observed they need to pass in an explicit number.

5 minutes seems like a fine default here, although I think I'm missing some context for the overall PR. Tying this sensor's frequency to the value set on `auto_observe_interval_minutes` would require that we also turn off any auto-observing functionality in the asset daemon to avoid two daemons trying to do the same job.

The machinery for auto-observing source assets is very different from the machinery for auto-materializing regular assets (for one, there's no system of rules for when source assets should be observed, it's just based on elapsed time), so in general I think it's reasonable to model this as just traditional sensor if we don't think there will be any need for more complex logic for when to kick these off (I'd buy that).

Some bits of trepidation:

1. we'd probably want to rename this property to something that's less similar to `auto_materialize`. as proposed, this is less declarative and "automatic" than the existing solution (as you need to add code in two places in order for your assets to actually be observed automatically). as you mention, this could be fixed by adding things to the Definitions machinery.
2. for users with tons of observable source assets, it might be offputting to end up with an equal number of sensors. this might be something that clever UI could handle (potentially grouping together all these auto-generated sensors).

Reply:
> for users with tons of observable source assets, it might be offputting to end up with an equal number of sensors. this might be something that clever UI could handle (potentially grouping together all these auto-generated sensors).

Absolutely. We'll want a version of this which can take multiple observable source assets. In fact @OwenKephart I think you should request changes here and then I should do exactly that.

> we'd probably want to rename this property to something that's less similar to auto_materialize. as proposed, this is less declarative and "automatic" than the existing solution (as you need to add code in two places in order for your assets to actually be observed automatically). as you mention, this could be fixed by adding things to the Definitions machinery.

We talked about this live, but the goal here is to provide a backwards compatibility layer. Agree that any new APIs should avoid the language.